cloudapi/logs.go

```diff
@@ -176,3 +176,186 @@ func (c *Config) StreamLogsToLogger(
+
+	// afterfn abstracts a timer
+	afterfn func(time.Duration) <-chan time.Time
+
+	// onRedialing is a callback invoked
+	// when a connection's refresh attempt is invoked.
+	onRedialing func()
+}
+
+type dialConfig struct {
+	url           string
+	headers       http.Header
+	retryAttempts int64
+	retryInterval time.Duration
+	refreshAfter  time.Duration
+}
+
+// OnRedialing sets a function to invoke
+// when the a new connection refresh is on going.
+func (wsc *wsconn) OnRedialing(f func()) {
+	wsc.onRedialing = f
+}
+
+func (wsc *wsconn) dial(ctx context.Context) error {
+	attempts := uint(1)
+	if wsc.dialConfig.retryAttempts > 0 {
+		attempts = uint(wsc.dialConfig.retryAttempts)
+	}
+	var newc LogtailStreamer
+	err := retry(attempts, wsc.dialConfig.retryInterval, func() (err error) {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			newc, err = wsc.dialer.DialContext(ctx, wsc.dialConfig.url, wsc.dialConfig.headers)
+		}
+		return
+	})
+	if err != nil {
+		return err
+	}
+
+	// if any, close the previous conn
+	if wsc.conn != nil {
+		_ = wsc.Close()
+	}
+	wsc.m.Lock()
+	wsc.conn = newc
+	wsc.m.Unlock()
+	return nil
+}
+
+// keepAlive keeps alive the websocket connection invoking a new dial
+// every interval defined from refreshAfter.
+func (wsc *wsconn) keepAlive(ctx context.Context, refreshAfter time.Duration) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-wsc.afterfn(refreshAfter):
+			wsc.onRedialing()
+			// error skipped to let it fail on reading
+			// or retry a new dial on the next timer firing
+			_ = wsc.dial(ctx)
+		}
+	}
+}
+
+// ReadMessge reads a message from the stream.
+func (wsc *wsconn) ReadMessage(ctx context.Context) ([]byte, error) {
+	readfn := func() (messageType int, p []byte, err error) {
+		wsc.m.Lock()
+		messageType, p, err = wsc.conn.ReadMessage()
+		wsc.m.Unlock()
+		return
+	}
+
+	isErr := func(err error) bool {
+		eclose := &websocket.CloseError{}
+		return err != nil && !errors.As(err, &eclose)
+	}
+
+	_, msg, err := readfn()
+	if isErr(err) {
+		// try to fix getting a fresh connection
+		wsc.onRedialing()
+		dErr := wsc.dial(ctx)
+		if dErr != nil {
+			// return the main error
+			return nil, err
+		}
+		_, msg, err = readfn()
+		if isErr(err) {
+			return nil, err
+		}
+	}
+	return msg, nil
+}
+
+// Close closes the connection.
+func (wsc *wsconn) Close() error {
+	_ = wsc.conn.WriteControl(
+		websocket.CloseMessage,
+		websocket.FormatCloseMessage(websocket.CloseGoingAway, "closing"),
+		time.Now().Add(time.Second))
+
+	wsc.m.Lock()
+	_ = wsc.conn.Close()
+	wsc.m.Unlock()
+	return nil
+}
+
+// retry retries a to exeucute a provided function until it isn't succeeded
+// or maximum number of attempts is hit. It waits the specificed interval
+// between the latest iteration and the next retry.
+func retry(attempts uint, interval time.Duration, do func() error) (err error) {
+	for i := 0; i < int(attempts); i++ {
+		if i > 0 {
+			time.Sleep(interval)
+		}
+		err = do()
```

Comment:
Could we do exponential backoff here to avoid flooding the backend with connection attempts?